File: 0020_valid_parentheses/valid_parentheses.cpp
#include <string>
#include <stack>
#include <unordered_map>

using namespace std;

bool is_valid(string s) {
    stack<char> stack;
    unordered_map<char, char> match = {
        {'(', ')'},
        {'[', ']'},
        {'{', '}'}
    };
    
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            stack.push(c);
        } else if (c == ')' || c == ']' || c == '}') {
            if (stack.empty()) {
                return false;
            }

            char top = stack.top();
            if (c != match[top]) {
                return false;
            }
            stack.pop();
        } else {
            return false;
        }
    }
    return stack.empty();
}

File: 0020_valid_parentheses/valid_parentheses.py
# Use a stack to keep track of brackets.
# Push opening brackets.
# If we find a matching closing bracket, we pop the opening bracket.
# At the end, we should have an empty stack.
def isValid(self, s: str) -> bool:
    stack = []
    match = {"(": ")", "[": "]", "{": "}"}
    for c in s:
        if c in "([{":
            stack.append(c)
        elif c in ")]}":
            if not stack:
                return False

            top = stack[-1]
            if c != match[top]:
                return False
            stack.pop()
        else:
            return False
    return not stack
